src/names/name-resolution.md

```diff
@@ -1,4 +1,348 @@
+
+r[names.resolution.expansion]
+## Expansion-time name resolution
+
+r[names.resolution.expansion.intro]
+Expansion-time name resolution is the stage of name resolution necessary to complete macro expansion and fully generate a crate's AST. This stage requires the resolution of macro invocations and use declarations. Resolving use declarations is required to resolve [path-based scope] macro invocations. Resolving macro invocations is required in order to expand them.
+
+After expansion-time name resolution, the AST must not contain any unexpanded macro invocations. Every macro invocation resolves to a valid definition that exists in the final AST (or an external crate). The resolution of imports must be *stable*. After expansion, imports in the fully expanded AST must resolve to the same definition, regardless of the order in which macros are expanded. Once the crate has been fully expanded all speculative import resolutions are validated to ensure that no new ambiguities were introduced by macro expansion.
+
+> [!NOTE]
+>
+> Due to the iterative nature of macro expansion, this causes so called time traveling ambiguities, such as when a glob import introduces an item that is ambiguous with its own base path.
+>
+> ```rust,compile_fail,E0659
+> macro_rules! m {
+>     () => { mod bar {} }
+> }
+>
+> mod bar {
+>     pub(crate) use m;
+> }
+>
+> fn f() {
+>     // * Initially speculatively resolve `bar` to the module in the crate root.
+>     // * Expansion of `m` introduces a second bar module inside the body of `f`.
+>     // * Expansion-time resolution finalizes resolutions by re-resolving all
+>     //   imports and macro invocations, sees the introduced ambiguity
+>     //   and reports it as an error.
+>     bar::m!(); // ERROR: `bar` is ambiguous
+> }
+> ```
+
+r[names.resolution.expansion.imports]
+### Imports
+
+
+r[names.resolution.expansion.imports.intro]
+All use declarations are fully resolved during this stage of resolution. Type-relative paths cannot be resolved at this stage of compilation and will produce an error.
+
+```rust,compile_fail,E0432
+mod my_mod {
+    pub const Const: () = ();
+
+    pub enum MyEnum {
+        MyVariant
+    }
+
+    impl MyEnum {
+        pub const Const: () = ();
+    }
+
+    pub type TypeAlias = MyEnum;
+}
+
+fn foo() {
+    // imports resolved at expansion-time
+    use my_mod::MyEnum; // OK
+    use my_mod::MyEnum::MyVariant; // OK
+    use my_mod::TypeAlias; // OK
+    use my_mod::TypeAlias::MyVariant; // Doesn't work
+    use my_mod::MyEnum::Const; // Doesn't work
+    use my_mod::Const; // OK
+
+    // expressions resolved during type-relative resolution
+    let _ = my_mod::TypeAlias::MyVariant; // OK
+    let _ = my_mod::MyEnum::Const; // OK
+}
+```
+
+r[names.resolution.expansion.imports.shadowing]
+The following is a list of situations where shadowing of use declarations is permitted:
+
+* [use glob shadowing]
+* [macro textual scope shadowing]
```

Comment:
Lists should usually start with a capital letter. Also, I realize we are not very consistent with this, but I think in this case the `use` would be inline code to differentiate it from the English word "use".

```suggestion
* [`use` glob shadowing]
* [Macro textual scope shadowing]
```